File: ios/RoutineChart/RoutineChart/Core/Utils/TokenGenerator.swift
//
//  TokenGenerator.swift
//  RoutineChart
//
//  Created for Phase 2.2: QR Family Joining
//

import Foundation
import CryptoKit

struct TokenGenerator {
    /// Generate a cryptographically secure random token
    /// - Parameter byteCount: Number of random bytes (default: 32)
    /// - Returns: Base64-encoded secure token string
    static func generateSecureToken(byteCount: Int = 32) -> String {
        var bytes = [UInt8](repeating: 0, count: byteCount)
        let result = SecRandomCopyBytes(kSecRandomDefault, byteCount, &bytes)
        
        guard result == errSecSuccess else {
            // Fallback to SymmetricKey if SecRandomCopyBytes fails
            let key = SymmetricKey(size: .bits256)
            return key.withUnsafeBytes { Data($0).base64EncodedString() }
        }
        
        return Data(bytes).base64EncodedString()
    }
    
    /// Generate a URL-safe token (alphanumeric only)
    /// - Parameter length: Desired token length (default: 32)
    /// - Returns: URL-safe token string
    static func generateURLSafeToken(length: Int = 32) -> String {
        let letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        var token = ""
        
        for _ in 0..<length {
            let randomIndex = Int.random(in: 0..<letters.count)
            let randomCharacter = letters[letters.index(letters.startIndex, offsetBy: randomIndex)]
            token.append(randomCharacter)
        }
        
        return token
    }
}

File: android/app/src/main/java/com/HammersTech/RoutineChart/core/utils/TokenGenerator.kt
package com.HammersTech.RoutineChart.core.utils

import android.util.Base64
import java.security.SecureRandom

/**
 * Token generator utility
 * Phase 2.2: QR Family Joining
 */
object TokenGenerator {
    /**
     * Generate a cryptographically secure random token
     * @param byteCount Number of random bytes (default: 32)
     * @return Base64-encoded secure token string
     */
    fun generateSecureToken(byteCount: Int = 32): String {
        val bytes = ByteArray(byteCount)
        SecureRandom().nextBytes(bytes)
        return Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_WRAP)
    }

    /**
     * Generate a URL-safe token (alphanumeric only)
     * @param length Desired token length (default: 32)
     * @return URL-safe token string
     */
    fun generateURLSafeToken(length: Int = 32): String {
        val chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        val random = SecureRandom()
        return (1..length)
            .map { chars[random.nextInt(chars.length)] }
            .joinToString("")
    }
}
